crates/libs/core/src/event.rs

```diff
@@ -1,252 +1,89 @@
 use super::*;
-use core::ffi::c_void;
-use core::marker::PhantomData;
-use core::mem::{size_of, transmute_copy};
-use core::ptr::null_mut;
-use std::sync::Mutex;
+use core::mem::transmute_copy;
+use std::sync::{Arc, RwLock};
 
 /// A type that you can use to declare and implement an event of a specified delegate type.
 ///
 /// The implementation is thread-safe and designed to avoid contention between events being
 /// raised and delegates being added or removed.
 pub struct Event<T: Interface> {
-    swap: Mutex<()>,
-    change: Mutex<()>,
-    delegates: Array<T>,
-}
-
-impl<T: Interface> Default for Event<T> {
-    fn default() -> Self {
-        Self::new()
-    }
+    delegates: RwLock<Option<Arc<[Delegate<T>]>>>,
 }
 
 impl<T: Interface> Event<T> {
     /// Creates a new, empty `Event<T>`.
     pub fn new() -> Self {
         Self {
-            delegates: Array::new(),
-            swap: Mutex::default(),
-            change: Mutex::default(),
+            delegates: RwLock::new(None),
         }
     }
 
     /// Registers a delegate with the event object.
-    pub fn add(&mut self, delegate: &T) -> Result<i64> {
-        let mut _lock_free_drop = Array::new();
-        Ok({
-            let _change_lock = self.change.lock().unwrap();
-            let mut new_delegates = Array::with_capacity(self.delegates.len() + 1)?;
-            for delegate in self.delegates.as_slice() {
-                new_delegates.push(delegate.clone());
-            }
-            let delegate = Delegate::new(delegate)?;
-            let token = delegate.to_token();
-            new_delegates.push(delegate);
+    pub fn add(&self, delegate: &T) -> Result<i64> {
+        let mut guard = self.delegates.write().unwrap();
+        let mut new_list = if let Some(old_delegates) = guard.as_ref() {
+            let mut new_list = Vec::with_capacity(old_delegates.len() + 1);
+            new_list.extend_from_slice(&old_delegates);
+            new_list
+        } else {
+            Vec::with_capacity(1)
```

Comment:
It would be ideal if we could create the `Arc` "with capacity" rather than via a temporary `Vec`. 🥲

Reply:
Note that `Arc` does provide a `FromIterator` implementation so we could `collect` directly into an `Arc` in some cases.

Reply:
Didn't see that before, thanks for the info 👍